File: LeetCode/1472-design-browser-history/Solution.java
class BrowserHistory {
    String[] map;
    int position;
    int len;

    public BrowserHistory(String homepage) {
        map = new String[5000];
        map[len = position = 0] = homepage;
    }
    
    public void visit(String url) {
        map[len = ++position] = url;
    }
    
    public String back(int steps) {
        return map[position = Math.max(0, position - steps)];
    }
    
    public String forward(int steps) {
        return map[position = Math.min(len, position + steps)];
    }
}

File: LeetCode/1472-design-browser-history/solution.c
#define MAX(x, y) (((x) > (y)) ? (x) : (y))
#define MIN(x, y) (((x) < (y)) ? (x) : (y))

typedef struct {
    char* map[5000];
    int pos;
    int len;
} BrowserHistory;

BrowserHistory* browserHistoryCreate(char * homepage) {
    BrowserHistory* obj = malloc(sizeof(*obj));
    obj->map[obj->len = obj->pos = 0] = homepage;
    return obj;
}

void browserHistoryVisit(BrowserHistory* obj, char * url) {
    obj->map[obj->len = ++(obj->pos)] = url;
}

char * browserHistoryBack(BrowserHistory* obj, int steps) {
    return obj->map[obj->pos = MAX(0, (obj->pos - steps))];
}

char * browserHistoryForward(BrowserHistory* obj, int steps) {
    return obj->map[obj->pos = MIN(obj->len, (obj->pos + steps))];
}

void browserHistoryFree(BrowserHistory* obj) {
    free(obj);
}